src/Pos/Types/Arbitrary.hs

```diff
@@ -68,9 +82,159 @@ derive makeArbitrary ''TxOut
+-- It is used in tests to make sure addition raises the appropriate exception when this
+-- happens.
+newtype CoinPairOverflowSum = TwoCoinsSum
+    { get2CSum :: (Coin, Coin)
+    } deriving (Show, Eq)
+
+instance Arbitrary CoinPairOverflowSum where
+    arbitrary = do
+        c1 <- arbitrary
+        let lowerBound = succ $ coinToInteger $ (maxBound @Coin) `unsafeSubCoin` c1
+            upperBound = coinToInteger (maxBound @Coin)
+        c2 <- mkCoin . fromIntegral <$> choose (lowerBound, upperBound)
+        return $ TwoCoinsSum (c1, c2)
+
+-- | This datatype has two coins that will never overflow when added.
+-- It is therefore safe to add them. Useful in tests to ensure adding two coins whose sum
+-- is a valid 'Coin' always works.
+newtype SafeCoinPairSum = CoinPairSum
+    { getPairSum :: (Coin, Coin)
+    } deriving (Show, Eq)
+
+instance Arbitrary SafeCoinPairSum where
+    arbitrary = do
+        c1 <- arbitrary
+        let upperBound = unsafeGetCoin c1
+            highestBound = unsafeGetCoin maxBound
+        c2 <- mkCoin <$> choose (0, highestBound - upperBound)
+        return $ CoinPairSum (c1, c2)
+
+-- | This datatype has two coins that will always underflow when subtracted.
+-- It is used in tests to make sure subtraction raises the appropriate exception when this
+-- happens.
+newtype CoinPairOverflowSub = TwoCoinsSub
+    { get2CSub :: (Coin, Coin)
+    } deriving (Show, Eq)
+
+instance Arbitrary CoinPairOverflowSub where
+    arbitrary = do
+        c2 <- arbitrary
+        let secondWord = unsafeGetCoin c2
+        c1 <- mkCoin <$> choose (0, secondWord - 1)
+        return $ TwoCoinsSub (c1, c2)
+
+-- | This datatype has two coins that will never underflow when subtracted.
+-- It is therefore safe to subtract them. Useful in tests to show that two coins whose
+-- subtraction does not underflow always works.
+newtype SafeCoinPairSub = CoinPairSub
+    { getPairSub :: (Coin, Coin)
+    } deriving (Show, Eq)
+
+instance Arbitrary SafeCoinPairSub where
+    arbitrary = do
+        c1 <- arbitrary
+        let upperBound = unsafeGetCoin c1
+        c2 <- mkCoin <$> choose (0, upperBound)
+        return $ CoinPairSub (c1, c2)
+
+-- | This datatype has a 'Coin' and an 'Integer' that will always overflow when
+-- multiplied.
+-- It is used in tests to make sure multiplication raises the appropriate exception when
+-- this happens.
+newtype CoinPairOverflowMul = TwoCoinsM
+    { get2CMul :: (Coin, Integer)
+    } deriving (Show, Eq)
+
+instance Arbitrary CoinPairOverflowMul where
+    arbitrary = do
+        c1 <- arbitrary
+        let integralC1 = coinToInteger c1
+            lowerBound =
+                integralC1 + (coinToInteger $ (maxBound @Coin) `divCoin` integralC1)
+            upperBound = coinToInteger (maxBound @Coin)
+        c2 <- fromIntegral @Integer <$> choose (lowerBound, upperBound)
+        return $ TwoCoinsM (c1, c2)
+
+-- | This datatype has two coins that will always overflow when multiplied.
```

Comment:
Thanks for pointing that out. I remember writing "A 'Coin' and an 'Integer'" somewhere, but I forgot I needed to write it twice (here and just before).